tests/data/test_artifact_dataset_sources.py

```diff
@@ -0,0 +1,121 @@
+):
+    dataset_source = resolve_dataset_source(source_uri)
+    assert isinstance(dataset_source, FileSystemDatasetSource)
+    assert dataset_source._get_source_type() == source_type
+    assert type(dataset_source).__name__ == source_class_name
+    assert type(dataset_source).__qualname__ == source_class_name
+    assert dataset_source.uri == source_uri
+
+
+@pytest.mark.parametrize(
+    ("source_uri", "source_type"),
+    [
+        ("/tmp/path/to/my/local/file.txt", "local"),
+        ("file:///tmp/path/to/my/local/directory", "local"),
+        ("s3://mybucket/path/to/my/file.txt", "s3"),
+        ("gs://mybucket/path/to/my/dir", "gs"),
+        ("wasbs://user@host.blob.core.windows.net/dir", "wasbs"),
+        ("ftp://mysite.com/path/to/my/file.txt", "ftp"),
+        ("sftp://mysite.com/path/to/my/dir", "sftp"),
+        ("hdfs://host_name:8020/hdfs/path/to/my/file.txt", "hdfs"),
+        ("viewfs://host_name:8020/path/to/my/dir", "viewfs"),
+    ],
+)
+def test_to_and_from_json(source_uri, source_type):
+    dataset_source = resolve_dataset_source(source_uri)
+    assert dataset_source._get_source_type() == source_type
+    source_json = dataset_source.to_json()
+
+    parsed_source_json = json.loads(source_json)
+    assert parsed_source_json["uri"] == source_uri
+
+    reloaded_source = get_dataset_source_from_json(
+        source_json, source_type=dataset_source._get_source_type()
+    )
+    assert isinstance(reloaded_source, FileSystemDatasetSource)
+    assert type(dataset_source) == type(reloaded_source)
+    assert reloaded_source.uri == dataset_source.uri
+
+
+@pytest.mark.parametrize(
+    ("source_uri", "source_type"),
+    [
+        ("/tmp/path/to/my/local/file.txt", "local"),
+        ("file:///tmp/path/to/my/local/directory", "local"),
+        ("s3://mybucket/path/to/my/file.txt", "s3"),
+        ("gs://mybucket/path/to/my/dir", "gs"),
+        ("wasbs://user@host.blob.core.windows.net/dir", "wasbs"),
+        ("ftp://mysite.com/path/to/my/file.txt", "ftp"),
+        ("sftp://mysite.com/path/to/my/dir", "sftp"),
+        ("hdfs://host_name:8020/hdfs/path/to/my/file.txt", "hdfs"),
+        ("viewfs://host_name:8020/path/to/my/dir", "viewfs"),
+    ],
+)
+def test_download_makes_expected_mlflow_artifacts_download_call(source_uri, source_type):
+    dataset_source = resolve_dataset_source(source_uri)
+    assert dataset_source._get_source_type() == source_type
+
+    with mock.patch("mlflow.data.artifact_dataset_sources.download_artifacts") as download_imp_mock:
+        dataset_source.download()
+        download_imp_mock.assert_called_once_with(source_uri)
+
+
+def test_local_downloads(tmp_path):
+    # Test string file paths
+    file_path = str(tmp_path / "myfile.txt")
+    with open(file_path, "w") as f:
+        f.write("text")
+
+    file_dataset_source = resolve_dataset_source(file_path)
+    assert file_dataset_source._get_source_type() == "local"
+    assert file_dataset_source.download() == file_path
```

Comment:
do we want to include a check here for the contexts of the file?

Reply:
I think file contents are a better check, you're right. We may adjust the returned path later on :).